### ISSUE TITLE
Multiple exceptions on `on_chdir` events when changing directory using z

### ISSUE THREAD

**Issue by nilehmann** [xonsh/xonsh#2504]
In my .xonshrc

```python
@events.on_chdir
def on_chdir(olddir, newdir, **kw):
     echo newdir
```
When I change the directory using z the following exception is raised infinitely

```bash
$ z Documents
xonsh: To log full traceback to a file set: $XONSH_TRACEBACK_LOGFILE = <filename>
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/xonsh/ptk/shortcuts.py", line 109, in prompt
    document = cli.run(reset_current_buffer=False)
  File "/usr/lib/python3.6/site-packages/prompt_toolkit/interface.py", line 411, in run
    with self.input.raw_mode():
  File "/usr/lib/python3.6/site-packages/prompt_toolkit/terminal/vt100_input.py", line 470, in __enter__
    termios.tcsetattr(self.fileno, termios.TCSANOW, newattr)
termios.error: (4, 'Interrupted system call')
xonsh: To log full traceback to a file set: $XONSH_TRACEBACK_LOGFILE = <filename>
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/xonsh/ptk/shortcuts.py", line 109, in prompt
    document = cli.run(reset_current_buffer=False)
  File "/usr/lib/python3.6/site-packages/prompt_toolkit/interface.py", line 411, in run
    with self.input.raw_mode():
  File "/usr/lib/python3.6/site-packages/prompt_toolkit/terminal/vt100_input.py", line 470, in __enter__
    termios.tcsetattr(self.fileno, termios.TCSANOW, newattr)
termios.error: (4, 'Interrupted system call')
...
```
My xonfig
```bash
$ xonfig 
+------------------+-----------------+
| xonsh            | 0.5.12          |
| Python           | 3.6.2           |
| PLY              | 3.9             |
| have readline    | True            |
| prompt toolkit   | 1.0.15          |
| shell type       | prompt_toolkit  |
| pygments         | 2.2.0           |
| on posix         | True            |
| on linux         | True            |
| distro           | arch            |
| on darwin        | False           |
| on windows       | False           |
| on cygwin        | False           |
| is superuser     | False           |
| default encoding | utf-8           |
| xonsh encoding   | utf-8           |
| encoding errors  | surrogateescape |
+------------------+-----------------+
```

**Comment by gforsyth** [comment]
Are you using the `z` xontrib? Or regular `z`? If you aren't using the xontrib, you should try it out: https://github.com/astronouth7303/xontrib-z
If you are, then we'll sort out if this is an issue with the xontrib or with events.

**Comment by nilehmann** [comment]
I'm using z xontrib.

**Comment by nilehmann** [comment]
I found that the following minimal example replicates the behavior

In my `.xonfigrc`
```
@events.on_chdir
def on_chdir(olddir, newdir, **kw):
     echo newdir
```

```bash
$ import xonsh.built_ins as built_ins
$ aliases['bug?'] = lambda:  built_ins.run_subproc([['cd', '~/Documents']])
$ bug?
~/Documents>
xonsh: To log full traceback to a file set: $XONSH_TRACEBACK_LOGFILE = <filename>
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/xonsh/ptk/shortcuts.py", line 109, in prompt
    document = cli.run(reset_current_buffer=False)
  File "/usr/lib/python3.6/site-packages/prompt_toolkit/interface.py", line 411, in run
    with self.input.raw_mode():
  File "/usr/lib/python3.6/site-packages/prompt_toolkit/terminal/vt100_input.py", line 470, in __enter__
    termios.tcsetattr(self.fileno, termios.TCSANOW, newattr)
termios.error: (5, 'Input/output error')
~/Documents>
xonsh: To log full traceback to a file set: $XONSH_TRACEBACK_LOGFILE = <filename>
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/xonsh/ptk/shortcuts.py", line 109, in prompt
    document = cli.run(reset_current_buffer=False)
  File "/usr/lib/python3.6/site-packages/prompt_toolkit/interface.py", line 411, in run
    with self.input.raw_mode():
  File "/usr/lib/python3.6/site-packages/prompt_toolkit/terminal/vt100_input.py", line 470, in __enter__
    termios.tcsetattr(self.fileno, termios.TCSANOW, newattr)
termios.error: (5, 'Input/output error')
.
.
.
```

It may not be a bug and perhaps I'm doing something wrong :S

**Comment by astronouth7303** [comment]
Do you still get an error if you use `print()` instead of `echo`?

@gforsyth This doesn't smell like a `z` problem. If anything, it's how events interact with command execution environment.

**Comment by nilehmann** [comment]
Yes, I still get an error, a different one but still.

```bash
~> import xonsh.built_ins as built_ins
~> import xonsh.built_ins as built_ins
~> @events.on_chdir
.. def on_chdir(olddir, newdir, **kwargs):
..     print(newdir)
..
~> aliases['bug?'] = lambda:  built_ins.run_subproc([['cd', '~/Documents']])
~> bug?
xonsh: To log full traceback to a file set: $XONSH_TRACEBACK_LOGFILE = <filename>
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 8487, in fire
    rv = handler(**kwargs)
  File "/home/nlehmann/.xonsh.d/rvm.xsh", line 3, in on_chdir
  File "/usr/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 12770, in write
    r = h.write(s)
  File "/usr/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 12770, in write
    r = h.write(s)
  File "/usr/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 12770, in write
    r = h.write(s)
  [Previous line repeated 978 more times]
  File "/usr/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 12768, in write
    h = self.handle
  File "/usr/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 12708, in handle
    return self.registry.get(threading.get_ident(), self.default)
RecursionError: maximum recursion depth exceeded in comparison

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 13063, in run
    r = self.f(self.args, sp_stdin, sp_stdout, sp_stderr, spec)
  File "/usr/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 12885, in proxy_two
    return f(args, stdin)
  File "/usr/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 10632, in cd
    _change_working_directory(d, follow_symlinks)
  File "/usr/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 10551, in _change_working_directory
    events.on_chdir.fire(olddir=old, newdir=env['PWD'])
  File "/usr/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 8489, in fire
    print_exception("Exception raised in event handler; ignored.")
  File "/usr/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 5983, in print_exception
    sys.stderr.write('xonsh: To log full traceback to a file set: '
  File "/usr/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 12770, in write
    r = h.write(s)
  File "/usr/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 12770, in write
    r = h.write(s)
  File "/usr/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 12770, in write
    r = h.write(s)
  [Previous line repeated 981 more times]
  File "/usr/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 12768, in write
    h = self.handle
  File "/usr/lib/python3.6/site-packages/xonsh/__amalgam__.py", line 12708, in handle
    return self.registry.get(threading.get_ident(), self.default)
RecursionError: maximum recursion depth exceeded in comparison
```

**Comment by nilehmann** [comment]
Ok, I found the issue. The problem occurs when a thread is spawn within a thread and the inner thread has been invoked with `captured=False`.

```python
# Stack overflow
built_ins.run_subproc([[
    lambda: built_ins.run_subproc([[lambda: print("foo")]],
    captured=false
)]])
```

```python
# No exception
built_ins.run_subproc([[
    lambda: built_ins.run_subproc([[lambda: print("foo")]],
    captured='stdout'  # or object or hiddenobject
)]])
```
The problem happens because at this [line](https://github.com/xonsh/xonsh/blob/f05c4d86b8529703832c2f0bb5fe2790dd3c5b66/xonsh/proc.py#L1347) `STDOUT_DISPATCHER` is equal to `sp_stdout` causing a circular reference.

I'm not sure whats the correct way to fix this.

**Comment by astronouth7303** [comment]
The fix probably gets caught up with #2337 but ummm I really really _really_ need to work on that. >.>

**Comment by nilehmann** [comment]
In the meantime I just patched `xontrib-z` to do `os.chdir(path)` instead of `run_subproc([['cd', path]])` which is working for me.

**Comment by astronouth7303** [comment]
_quickly grabs xontrib maintainer hat_

@nilehmann I would merge that in if you used `xonsh.dirstack.cd()` instead of `os.chdir()`.

**Comment by nilehmann** [comment]
Cool, I'll make a PR tonight.

**Comment by astronouth7303** [comment]
(The reason for that use is compatibility with `autoxsh`, but I'm comfortable dropping that at this point.)

**Comment by scopatz** [comment]
Is this closeable?

**Comment by astronouth7303** [comment]
tbh, probably not? Not being able to run subcommands in event handlers is kind of a big gotcha for a shell.

**Comment by scopatz** [comment]
Oh, for some reason thought this was fixed by astronouth7303/xontrib-z#6. I agree we need to be able to run commands in the event handler.

**Comment by astronouth7303** [comment]
So I think the root cause was that an alias (a thread) triggered an event, and that event handler isn't in a good place.

The work-around is to link the event system with the command loop so that they are only handled on the main loop between commands (which isn't great, and I suspect somewhat fragile).

A better option would be to actual fix how we do some things, but that would involve tracking what threads and bits of code have what IO environments and honestly, that sounds like a messy gordian knot at the bottom of a can of worms.
